File: lca/src/main/java/com/dreampany/lca/data/source/api/CoinDataSource.java
package com.dreampany.lca.data.source.api;

import com.dreampany.frame.data.source.api.DataSource;
import com.dreampany.lca.data.enums.CoinSource;
import com.dreampany.lca.data.model.Coin;
import com.dreampany.lca.data.enums.Currency;

import io.reactivex.Maybe;

import java.util.List;

/**
 * Created by Hawladar Roman on 29/5/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 */
public interface CoinDataSource extends DataSource<Coin> {

    List<Coin> getItems(CoinSource source, Currency currency, int index, int limit);

    Maybe<List<Coin>> getItemsRx(CoinSource source, Currency currency, int index, int limit);

    Coin getItem(CoinSource source, Currency currency, long coinId);

    Maybe<Coin> getItemRx(CoinSource source, Currency currency, long coinId);

    List<Coin> getItems(CoinSource source, Currency currency, List<Long> coinIds);

    Maybe<List<Coin>> getItemsRx(CoinSource source, Currency currency, List<Long> coinIds);
}

File: tools/src/main/kotlin/com/dreampany/tools/data/source/api/CoinDataSource.kt
package com.dreampany.tools.data.source.api

import com.dreampany.framework.data.source.api.DataSource
import com.dreampany.tools.data.enums.CoinSort
import com.dreampany.tools.data.enums.Currency
import com.dreampany.tools.data.enums.Order
import com.dreampany.tools.data.model.Coin
import io.reactivex.Maybe

/**
 * Created by roman on 2019-11-11
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
interface CoinDataSource : DataSource<Coin> {

    fun getItem(
        currency: Currency,
        id: String
    ): Coin

    fun getItems(
        currency: Currency,
        sort: CoinSort,
        order: Order,
        start: Long,
        limit: Long
    ): List<Coin>?

    fun getItemsRx(
        currency: Currency,
        sort: CoinSort,
        order: Order,
        start: Long,
        limit: Long
    ): Maybe<List<Coin>>

    fun getItems(
        currency: Currency,
        ids: List<String>
    ): List<Coin>?

    fun getItemsRx(
        currency: Currency,
        ids: List<String>
    ): Maybe<List<Coin>>
}